service/controller/clusterapi/v28/machine_deployment_resource_set.go

```diff
@@ -83,9 +87,26 @@ func NewMachineDeploymentResourceSet(config MachineDeploymentResourceSetConfig)
 	}
 
 	initCtxFunc := func(ctx context.Context, obj interface{}) (context.Context, error) {
+		var cr v1alpha1.Cluster
+		{
+			md, err := key.ToMachineDeployment(obj)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+
+			id := key.WorkerClusterID(md)
```

Comment:
Is this controller watching MachineDeployments? We usually were naming controllers after object type they were watching so I was pretty convinced it was watching cluster objects so far. But this makes sense. Do we have some diagrams showing relationships between the controllers and the Cluster API objects?

Reply:
We named the controller `clusterapi`. That is not an object. There is probably some cleanup necessary at some point but we are still in "make it work". The picture below may be the best we have in terms of diagrams. Naming is already a bit outdated but it should transport the idea. 

<img width="1349" alt="Screenshot 2019-06-12 at 15 01 39" src="https://user-images.githubusercontent.com/552769/59353343-080dba80-8d23-11e9-8baf-45990c724d8e.png">

Reply:
I understand thing modified here is the "NodePoolController" and we don't have "ClusterController" yet. Correct?

Reply:
We have both controllers already. 

* https://github.com/giantswarm/aws-operator/blob/master/service/controller/clusterapi/cluster.go
* https://github.com/giantswarm/aws-operator/blob/master/service/controller/clusterapi/machine_deployment.go

They are just not in their own directory. We have three different controllers now in `clusterapi`. Two of them watch `Cluster` types. One of them watches `MachineDeployment` types.

Reply:
That doesn't look right to me. So far we had controllers in separate directories. What is the reason to keep them together? Is it a short term solution?

Reply:
> There is probably some cleanup necessary at some point but we are still in "make it work".

The reason is the transition we are in. We should get this straight at some point.